### ISSUE TITLE
sqlalchemy support

### ISSUE THREAD

**Issue by thomaspoignant** [sqlfluff/sqlfluff#1216]
I am using [sqlalchemy](https://www.sqlalchemy.org/) on my projects and the syntax is not valid for `sqlfluff`.
It would be great to have a templater for sqlalchemy to recognize the syntax.

Currently running it results with **unparsable errors**.

### Example
```sql
-- get_queues.sql
SELECT
    queue_id,
    queue_name,
    icon,
    code,
    type,
    suggested_reasons,
    COALESCE(ad_count, 0) AS ad_count
FROM
    queues
LEFT JOIN (
    SELECT
        UNNEST(queue_ids) AS id,
        COUNT(*) AS ad_count
    FROM
        moderation_inbox_items
    WHERE (:assigned_entity)::text IS NULL OR assigned_entity = (:assigned_entity)::text
    GROUP BY
        UNNEST(queue_ids)
)
ON id = queue_id
ORDER BY rank, queue_id
```
Is failing because of `:assigned_entity`

**Comment by alanmcruickshank** [comment]
This shouldn't be too difficult. I've got one thing I'm not sure about though. Do you know how sqlalchemy knows not to count ``::text` (which ends with `:text`) as a variable?

**Comment by jacopofar** [comment]
I'd like to try and implement it, and it seems now the best way is to create a plugin for it. I read the code of the DBT one but is not totally clear to me how it interfaces to SQLFluff (I don't know DBT).

There's a [`process`](https://github.com/sqlfluff/sqlfluff/blob/87197306a417a99aaccae62afacda3ac428f5931/plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py#L434) method to parse the file and I see it returns an instance of `[TemplatedFile](https://github.com/sqlfluff/sqlfluff/blob/87197306a417a99aaccae62afacda3ac428f5931/src/sqlfluff/core/templaters/base.py#L64)` which if I understand correctly maintains the reference to the original code so that later it can be "un-templated", is it correct?

**Comment by pwildenhain** [comment]
@jacopofar Have you ever used the pluggy package? That is what we use to manage our plugins -- you can see what we make "pluggable" by checking out the src/sqfluff/core/plugins directory. Right now it's just custom rules and templaters. And then I suggest reading up on pluggy to see how you can plug in the sqlalchemy templater. I can guarantee you won't need anywhere near as much complexity as the dbt templater 😉

**Comment by jacopofar** [comment]
I made a draft of it, https://github.com/jacopofar/sqlfluff/blob/add-sqlalchemy-plugin/plugins/sqlfluff-templater-sqlalchemy/sqlfluff_templater_sqlalchemy/templater.py more tests and maybe docs are needed but I think the logic is mostly there.

I wonder now if it shouldn't rather be a built-in templater instead, since there are no dependencies and SQLAlchemy is quite ubiquous. Similarly I would add also Psycopg2 (and 3), [they also use a regex](https://github.com/psycopg/psycopg/blob/31f1dc8b558459ac1720c4a86aff46eb71628b6c/psycopg/psycopg/_queries.py#L215) so in both cases no dependencies would be needed and it would cover two common libraries. It would also possible to cover [all param styles](https://www.python.org/dev/peps/pep-0249/#paramstyle) later on the same way.

A problem I am facing is how the user is supposed to define the parameters: the values from the config files are always `str`, but when one uses SQLAlchemy the values are quoted or not based on the Python type. So for example:

    SELECT * from bla WHERE id = :myid

will be replaced with `'42'` or `42` depending on the type of `myid`. Trying to guess it here (e.g. if it can be converted to float then do not quote, if it is within square brackets generate a list) seems dangerous and also very complex. I suggest to ask the user to insert literal strings that will be replaced verbatim when templating, so they have to write `'42'`, `(42)` or `42` in the template depending on the type.

**Comment by pwildenhain** [comment]
Great work so far 😎

> I wonder now if it shouldn't rather be a built-in templater instead, since there are no dependencies and SQLAlchemy is quite ubiquous.

🤔

This thinking makes sense. While it would be tidier to keep all templaters as plug-ins, for maintainability sake it really only matters if they're introducing other dependencies. And the jinja templaters is already built-in anyway

> I suggest to ask the user to insert literal strings that will be replaced verbatim when templating

AFAIK this is what we have users do now when using the jinja templater -- define the variables in their .sqlfluff file. So if I'm understanding correctly, I think that should work

**Comment by jacopofar** [comment]
Great! Then I'll rewrite it as an embedded templater (should be the same logic, just not as a plugin) and write more tests for a PR. Once it's in good shape should be trivial to copy it to another templater for the psycopg parameter style, it is just a different regex, but that would be a separate PR
